packages/cli/src/util/dev/server.ts

```diff
@@ -689,17 +690,39 @@ export default class DevServer {
     }
 
     // legacy NOW_REGION env variable
-    runEnv['NOW_REGION'] = 'dev1';
     buildEnv['NOW_REGION'] = 'dev1';
-    allEnv['NOW_REGION'] = 'dev1';
 
-    // mirror how VERCEL_REGION is injected in prod/preview
-    // only inject in `runEnvs`, because `allEnvs` is exposed to dev command
-    // and should not contain VERCEL_REGION
     if (this.projectSettings?.autoExposeSystemEnvs) {
+      // mirror how VERCEL_REGION is injected in prod/preview
+      // only inject in `runEnvs`, because `allEnvs` is exposed to dev command
+      // and should not contain VERCEL_REGION
       runEnv['VERCEL_REGION'] = 'dev1';
+
+      // expose VERCEL_ENV and VERCEL_URL to dev command
+      buildEnv['VERCEL_ENV'] = 'development';
+      if (this.address.host) {
+        buildEnv['VERCEL_URL'] = this.address.host;
+      }
+
+      // automaticaly expose VERCEL_URL, VERCEL_ENV
+      // as prefixed (framework) env vars
+      if (this.projectSettings?.framework) {
+        const frameworkSlug = this.projectSettings.framework;
+        const framework = frameworkList.find(f => f.slug === frameworkSlug);
+        const envPrefix = framework?.envPrefix;
+
+        if (envPrefix) {
+          buildEnv = {
+            ...buildEnv,
+            ...getPrefixedEnvVars({ envPrefix: envPrefix, envs: buildEnv }),
+          };
+        }
+      }
     }
 
+    // re-compute allEnv with altered envs above
+    allEnv = { ...buildEnv, ...runEnv };
```

Comment:
@styfle  from the old code comment above, it sounds like I should not include `runEnv` here. What do you think?
I'm referring to this comment:

>       // mirror how VERCEL_REGION is injected in prod/preview
>       // only inject in `runEnvs`, because `allEnvs` is exposed to dev command
>       // and should not contain VERCEL_REGION

Reply:
You shouldn't need to recompute. Instead, move the code in the `if` statement up before the `allEnv` is created.